akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress
+
+import scala.collection.immutable
+
+// TODO describe / think about how we could handle an actor dying and removing from "top"
+// TODO this is the scenario when "super traffic actor" goes away and we want to not use up the slot in HH
+
+/**
+ * INTERNAL API
+ *
+ * Mutable.
+ *
+ * Keeps a number of specific heavy hitters around in memory.
+ *
+ * See also Section 5.2 of http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf
+ * for a discussion about the assumptions made and guarantees about the Heavy Hitters made in this model.
+ * We assume the Cash Register model in which there are only additions, which simplifies HH detecion significantly.
+ */
+private[remote] sealed class HeavyHitters[T](strategy: HeavyHittersStrategy[T]) {
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
+
+  def items: immutable.Set[T] = _items.map(_.item)(scala.collection.breakOut).toSet // TODO avoid the toSet? does anyone care?
+
+  /**
+   * Attempt adding item to heavy hitters set, if it does not fit in the top yet,
+   * it will be dropped and the method will return `false`.
+   *
+   * @return `true` if the added item is indeed a heavy hitter
+   */
+  def update(item: T, count: Long): Boolean =
+    _items.find(_.item == item) match { // TODO optimise? could it be the same case always?
```

Comment:
Thinking about it even more, there is probably not much sense in a priority queue either, since removal of the smallest will be less frequent than updating values. Sorted arrays hence make more sense, since then `find` is just a  binary search. Removal then will make a hole, which you need to fix by `System.arrayCopy`. That should be also fast.